src/Equinor.ProCoSys.Auth/Authentication/ApiAuthenticator.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Identity.Client;
+
+namespace Equinor.ProCoSys.Auth.Authentication
+{
+    /// <summary>
+    /// Abstract class to be used for acquire a token to authentice a HttpClient to access a 
+    /// foreign API. The authentication can be done in two ways via AuthenticationType:
+    ///  * on behalf of logged on user using the OAuth 2.0 On-Behalf-Of flow
+    ///  * as the client itself (in the name of no user) using the client credentials flow
+    /// </summary>
+    public abstract class ApiAuthenticator : IBearerTokenSetter
+    {
+        protected readonly ILogger<ApiAuthenticator> _logger;
+
+        private readonly IAuthenticatorOptions _options;
+        private readonly string _secretInfo;
+        private readonly string _apiScope;
+        private string _bearerTokenFromRequest;
+
+        /// <summary>
+        /// cache created OnBehalfOf tokens during this request
+        /// </summary>
+        private readonly ConcurrentDictionary<string, string> _oboTokenCache
+            = new ConcurrentDictionary<string, string>();
+        /// <summary>
+        /// cache created application client tokens during this request
+        /// </summary>
+        private readonly ConcurrentDictionary<string, string> _appTokenCache
+            = new ConcurrentDictionary<string, string>();
+
+        public ApiAuthenticator(string apiScopeKey, IAuthenticatorOptions options, ILogger<ApiAuthenticator> logger)
+        {
+            if (!options.Scopes.TryGetValue(apiScopeKey, out _apiScope))
+            {
+                throw new ArgumentException(
+                    $"List of scopes in {nameof(IAuthenticatorOptions)} dont have key {apiScopeKey}. " + 
+                    $"Check {nameof(IAuthenticatorOptions)} implementation when filling {nameof(options.Scopes)} from the application config");
+            }
+            _options = options;
+            _logger = logger;
+            var secret = _options.Secret;
+            _secretInfo = $"{secret.Substring(0, 2)}***{secret.Substring(secret.Length - 1, 1)}";
+        }
+
+        public AuthenticationType AuthenticationType { get; set; }
```

Comment:
Curious: Where in the code is AuthenticationType set to OnBehalfOf?

Reply:
Probably because AuthenticationType is not explicit set, just have the default value for an enum which coincidentallyis  OnBehalfOf.

I'll set it explicit in ApiAuthenticator.cs ctor. OK @jsoi-eq ?